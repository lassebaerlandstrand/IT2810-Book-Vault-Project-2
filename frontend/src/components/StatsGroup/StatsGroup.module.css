.root {
  display: flex;
  background-image: linear-gradient(
    -60deg,
    var(--mantine-color-blue-4) 0%,
    var(--mantine-color-blue-7) 100%
  );
  padding: calc(var(--mantine-spacing-xl) * 1.5);
  border-radius: var(--mantine-radius-md);
}

.title {
  color: var(--mantine-color-white);
  text-transform: uppercase;
  font-weight: 700;
}

.count {
  color: var(--mantine-color-white);
  font-size: 32px;
  font-weight: 700;
}

.description {
  color: var(--mantine-color-blue-0);
  font-size: var(--mantine-font-size-sm);
}

.stat {
  flex: 1;

  & + & {
    padding-left: var(--mantine-spacing-xl);
    margin-left: var(--mantine-spacing-xl);
    border-left: 1px solid var(--mantine-color-blue-3);
  }
}
