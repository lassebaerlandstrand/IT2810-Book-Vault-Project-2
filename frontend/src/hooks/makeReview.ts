import { useMutation } from '@apollo/client';
import { CreateReviewInput } from '@/generated/graphql';
import { CREATE_REVIEW } from '@/graphql/mutations/reviews';

/**
 * Custom hook that handles review creation functionality.
 * @returns Object containing:
 * - submitReview: Function to create a new review
 * - loading: Boolean indicating if the mutation is in progress
 * - error: Any error that occurred during the mutation
 */
export const makeReview = () => {
  // Return the mutation and its states (data, loading, error)
  const [createReview, { data, loading, error }] = useMutation(CREATE_REVIEW);

  const submitReview = async (input: CreateReviewInput) => {
    try {
      await createReview({
        variables: { input },
        update: (cache, { data: mutationData }) => {
          if (mutationData?.createReview.success) {
            // On update remove the bookReviews "focusUserUUID" queries from cache as
            // they are now stale
            const allKeys = cache.extract().ROOT_QUERY;

            Object.keys(allKeys).forEach((key) => {
              if (key.startsWith('bookReviews({"focusUserUUID"')) {
                const jsonString = key.replace('bookReviews(', '').replace(')', '');
                const parsed = JSON.parse(jsonString);
                const { limit, offset } = parsed;

                cache.evict({
                  id: 'ROOT_QUERY',
                  fieldName: 'bookReviews',
                  args: {
                    focusUserUUID: input.userUUID,
                    limit,
                    offset,
                  },
                });
              }
            });

            cache.gc();
          }
        },
      });
    } catch (e) {
      console.error('Error during update:', e);
    }
  };

  return {
    submitReview, // Function to trigger the mutation later
    success: data?.createReview.success ?? !error,
    message: data?.createReview.message ?? (error && 'Error making review. Try again later.'),
    loading,
    error,
  };
};
