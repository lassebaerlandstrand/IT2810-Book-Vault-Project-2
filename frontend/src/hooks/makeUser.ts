import { useMutation } from '@apollo/client';
import { CREATE_USER } from '@/graphql/mutations/users';

/**
 * Custom hook that creates a new user.
 * @returns Object containing:
 * - user: The created user data or null
 * - loading: Boolean indicating if the mutation is in progress
 * - error: Any error that occurred during the mutation
 */
export const makeUser = () => {
  const [createUser, { data, loading, error }] = useMutation(CREATE_USER);

  // useEffect(() => {
  //   const createUserAsync = async () => {
  //     try {
  //       await createUser();
  //     } catch (e) {
  //       console.error('Error during user creation:', e);
  //     }
  //   };

  //   createUserAsync();
  // }, [createUser]);

  return {
    createUser,
    user: data?.createUser || null,
    loading,
    error,
  };
};
