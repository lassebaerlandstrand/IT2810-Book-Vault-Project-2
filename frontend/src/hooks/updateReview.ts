//UPDATE_REVIEW

import { useMutation } from '@apollo/client';
import { UpdateReviewInput } from '@/generated/graphql';
import { UPDATE_REVIEW } from '@/graphql/mutations/reviews';

export const updateReview = () => {
  const [update, { data, loading, error }] = useMutation(UPDATE_REVIEW, {
    update(cache) {
      // On update remove the bookReviews "focusUserUUID" queries from cache as
      // they are now stale
      const allKeys = cache.extract().ROOT_QUERY;

      Object.keys(allKeys).forEach((key) => {
        if (key.startsWith('bookReviews({"focusUserUUID"')) {
          cache.evict({
            id: cache.identify({
              __typename: 'Query',
              key,
            }),
          });
        }
      });

      //Clean up
      cache.gc();
    },
  });

  const submitUpdate = async (input: UpdateReviewInput) => {
    try {
      await update({
        variables: { input },
      });
    } catch (e) {
      console.error('Error during update:', e);
    }
  };

  return {
    submitUpdate,
    success: data?.updateReview.success,
    message: data?.updateReview.message,
    loading,
    error,
  };
};
