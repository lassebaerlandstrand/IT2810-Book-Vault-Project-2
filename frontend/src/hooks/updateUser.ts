import { useMutation } from '@apollo/client';
import { UpdateUserInput } from '@/generated/graphql';
import { UPDATE_USER } from '@/graphql/mutations/users';

export const updateUser = () => {
  const [updateUser, { loading, error }] = useMutation(UPDATE_USER);

  const submitUpdate = async (input: UpdateUserInput) => {
    try {
      await updateUser({
        variables: { input },
      });
    } catch (e) {
      console.error('Error during user update:', e);
    }
  };

  return {
    submitUpdate,
    loading,
    error,
  };
};
