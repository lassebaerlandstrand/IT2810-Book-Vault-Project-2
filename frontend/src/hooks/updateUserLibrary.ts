import { Reference, useMutation } from '@apollo/client';
import { UpdateUserLibraryInput } from '@/generated/graphql';
import { UPDATE_USERLIBARY } from '@/graphql/mutations/users';

export const updateUserLibrary = () => {
  const [UpdateUserLibrary, { data, loading, error }] = useMutation(UPDATE_USERLIBARY);

  const submitUpdate = async (input: UpdateUserLibraryInput) => {
    try {
      await UpdateUserLibrary({
        variables: { input },
        update: (cache, { data: mutationData }) => {
          if (mutationData?.updateUserLibrary.success) {
            const allKeys = cache.extract().ROOT_QUERY;

            // As Apollo Client doesnt support wildcards in arguments for identifying queries,
            // we have to filter through the cache like it was text looking for relevant
            // queries to evict

            Object.keys(allKeys).forEach((key) => {
              if (key.startsWith('books(')) {
                const jsonString = key.replace('books(', '').replace(')', '');
                const parsed = JSON.parse(jsonString);
                const { sortBy, sortOrder } = parsed.input.sortInput;
                const { limit, offset } = parsed;

                if (key.includes(`"haveReadListUserUUID":"${input.UUID}"`)) {
                  cache.evict({
                    id: 'ROOT_QUERY',
                    fieldName: 'books',
                    args: {
                      input: {
                        haveReadListUserUUID: input.UUID,
                        sortInput: { sortBy, sortOrder },
                      },
                      limit,
                      offset,
                    },
                  });
                } else if (key.includes(`"wantToReadListUserUUID":"${input.UUID}"`)) {
                  cache.evict({
                    id: 'ROOT_QUERY',
                    fieldName: 'books',
                    args: {
                      input: {
                        wantToReadListUserUUID: input.UUID,
                        sortInput: { sortBy, sortOrder },
                      },
                      limit,
                      offset,
                    },
                  });
                }
              }
            });

            cache.modify({
              id: cache.identify({ __typename: 'User', UUID: input.UUID }),
              fields: {
                wantToRead(existing = [], { toReference, readField }) {
                  const filtered = existing.filter(
                    (bookRef: Reference) =>
                      readField('id', bookRef) !== input.haveRead &&
                      readField('id', bookRef) !== input.removeFromLibrary
                  );

                  if (input.wantToRead) {
                    const newBookRef = toReference({ __typename: 'Book', id: input.wantToRead });
                    return [...filtered, newBookRef];
                  }

                  return filtered;
                },
                haveRead(existing = [], { toReference, readField }) {
                  const filtered = existing.filter(
                    (bookRef: Reference) =>
                      readField('id', bookRef) !== input.wantToRead &&
                      readField('id', bookRef) !== input.removeFromLibrary
                  );

                  if (input.haveRead) {
                    const newBookRef = toReference({ __typename: 'Book', id: input.haveRead });
                    return [...filtered, newBookRef];
                  }

                  return filtered;
                },
              },
            });
          }
        },
      });
    } catch (e) {
      console.error('Error during user update:', e);
    }
  };

  return {
    submitUpdate,
    success: data?.updateUserLibrary.success ?? !error,
    message: data?.updateUserLibrary.message ?? (error && 'Error updating user. Try again later.'),
    loading,
    error,
  };
};

/*
// Throw out old
const allKeys = cache.extract().ROOT_QUERY;

Object.keys(allKeys).forEach((key) => {
  
  
  if (key.startsWith(`books({"input":{"haveReadListUserUUID":"${input.UUID}"`) || key.startsWith(`books({"input":{"wantToReadListUserUUID":"${input.UUID}"`)) {
    cache.evict({
      id: cache.identify({
        __typename: 'Query',
        key,
      }),
    });
  }
  
});
*/
