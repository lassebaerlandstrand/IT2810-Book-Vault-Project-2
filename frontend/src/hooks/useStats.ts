import { useQuery } from '@apollo/client';
import { GET_STATS } from '@/graphql/queries/stats';

/**
 * Custom hook that fetches statistics about the website. Gets the total number of books, authors, and ratings.
 * @returns Object containing:
 * - stats: Object containing the statistics
 * - loading: Boolean indicating if the query is in progress
 * - error: Any error that occurred during the query
 */
export const useStats = () => {
  const { data, loading, error, refetch } = useQuery(GET_STATS);

  return {
    stats: data?.stats,
    loading,
    error,
    refetch,
  };
};
