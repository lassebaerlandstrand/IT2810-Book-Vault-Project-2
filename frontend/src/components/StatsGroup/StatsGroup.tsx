import { ApolloError } from '@apollo/client/errors';
import { Box, Container, Text } from '@mantine/core';
import { Stats } from '@/generated/graphql';
import { useStats } from '@/hooks/useStats';
import { formatNumberWithSpaces } from '@/utils/formatting';
import styles from './StatsGroup.module.css';

type StatsType = Omit<Stats, '__typename'>;

const getField = (
  stats: StatsType | undefined,
  field: keyof StatsType,
  loading: boolean,
  error: ApolloError | undefined
): string => {
  if (loading) {
    return '. . .';
  }
  if (error || stats == null) {
    return '-';
  }
  return formatNumberWithSpaces(stats[field]);
};

export const StatsGroup = () => {
  const { stats, loading, error } = useStats();

  const data = [
    {
      title: 'Books in our collection',
      stats: getField(stats, 'totalBooks', loading, error),
      description:
        'Explore our vast collection of books, where every title opens the door to new adventures and ideas!',
    },
    {
      title: 'Authors featured',
      stats: getField(stats, 'totalAuthors', loading, error),
      description:
        'Immerse yourself in stories from a diverse set of authors, each bringing a unique perspective to our library.',
    },
    {
      title: 'Number of ratings',
      stats: getField(stats, 'totalRatings', loading, error),
      description:
        'With a massive collection of ratings, you have the insights you need to discover your next must-read.',
    },
  ];

  return (
    <Box className={styles.wrapper}>
      {data.map((stat, i) => (
        <Container key={i} className={styles.stat} aria-live="polite">
          <Text className={styles.count}>{stat.stats}</Text>
          <Text className={styles.title}>{stat.title}</Text>
          <Text className={styles.description}>{stat.description}</Text>
        </Container>
      ))}
    </Box>
  );
};
