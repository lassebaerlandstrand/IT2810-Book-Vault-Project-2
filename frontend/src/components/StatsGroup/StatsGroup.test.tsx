import { render, screen } from '@test-utils';
import { useStats } from '@/hooks/useStats';
import { StatsGroup } from './StatsGroup';

vi.mock('@/hooks/useStats');

describe('StatsGroup component', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders correctly', () => {
    (useStats as jest.Mock).mockReturnValue({
      stats: {
        totalBooks: 1000,
        totalAuthors: 200,
        totalRatings: 5000,
      },
      loading: false,
      error: null,
      refetch: vi.fn(),
    });

    const { asFragment } = render(<StatsGroup />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('shows loading state correctly', () => {
    (useStats as jest.Mock).mockReturnValue({
      stats: null,
      loading: true,
      error: null,
      refetch: vi.fn(),
    });

    render(<StatsGroup />);

    expect(screen.getAllByText('. . .')).toHaveLength(3);
  });

  it('handles error state correctly', () => {
    (useStats as jest.Mock).mockReturnValue({
      stats: null,
      loading: false,
      error: new Error('Error'),
      refetch: vi.fn(),
    });

    render(<StatsGroup />);

    expect(screen.getAllByText('-')).toHaveLength(3);
  });

  it('displays the correct stats', () => {
    (useStats as jest.Mock).mockReturnValue({
      stats: {
        totalBooks: 1000,
        totalAuthors: 200,
        totalRatings: 5000,
      },
      loading: false,
      error: null,
      refetch: vi.fn(),
    });

    render(<StatsGroup />);

    expect(screen.getByText('1 000')).toBeInTheDocument();
    expect(screen.getByText('200')).toBeInTheDocument();
    expect(screen.getByText('5 000')).toBeInTheDocument();
  });
});
