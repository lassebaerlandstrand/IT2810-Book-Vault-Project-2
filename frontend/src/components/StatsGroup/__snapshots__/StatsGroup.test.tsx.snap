// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StatsGroup component > renders correctly 1`] = `
<DocumentFragment>
  <style
    data-mantine-styles="true"
  >
    :root{--mantine-font-family: Poppins, -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji;--mantine-font-family-headings: Poppins, -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji;--mantine-breakpoint-xs: 30em;--mantine-breakpoint-md: 64em;--mantine-breakpoint-xl: 90em;--mantine-breakpoint-xxs: 24em;--mantine-color-dark-blue-0: #155896;--mantine-color-dark-blue-1: #124b81;--mantine-color-dark-blue-2: #0f3f6c;--mantine-color-dark-blue-3: #0c3256;--mantine-color-dark-blue-4: #092641;--mantine-color-dark-blue-5: #06192b;--mantine-color-dark-blue-6: #051321;--mantine-color-dark-blue-7: #030d16;--mantine-color-dark-blue-8: #030a11;--mantine-color-dark-blue-9: #02070b;--mantine-color-beige-0: #fcfaf7;--mantine-color-beige-1: #f9f5ef;--mantine-color-beige-2: #f6f0e7;--mantine-color-beige-3: #f2eade;--mantine-color-beige-4: #eee4d3;--mantine-color-beige-5: #e9ddc8;--mantine-color-beige-6: #e2d1b5;--mantine-color-beige-7: #dac5a1;--mantine-color-beige-8: #d4bd94;--mantine-color-beige-9: #6a5f4a;}:root[data-mantine-color-scheme="dark"]{--mantine-color-dark-blue-text: var(--mantine-color-dark-blue-4);--mantine-color-dark-blue-filled: var(--mantine-color-dark-blue-8);--mantine-color-dark-blue-filled-hover: var(--mantine-color-dark-blue-9);--mantine-color-dark-blue-light: rgba(5, 19, 33, 0.15);--mantine-color-dark-blue-light-hover: rgba(5, 19, 33, 0.2);--mantine-color-dark-blue-light-color: var(--mantine-color-dark-blue-3);--mantine-color-dark-blue-outline: var(--mantine-color-dark-blue-4);--mantine-color-dark-blue-outline-hover: rgba(9, 38, 65, 0.05);--mantine-color-beige-text: var(--mantine-color-beige-4);--mantine-color-beige-filled: var(--mantine-color-beige-8);--mantine-color-beige-filled-hover: var(--mantine-color-beige-9);--mantine-color-beige-light: rgba(226, 209, 181, 0.15);--mantine-color-beige-light-hover: rgba(226, 209, 181, 0.2);--mantine-color-beige-light-color: var(--mantine-color-beige-3);--mantine-color-beige-outline: var(--mantine-color-beige-4);--mantine-color-beige-outline-hover: rgba(238, 228, 211, 0.05);}:root[data-mantine-color-scheme="light"]{--mantine-color-dark-blue-text: var(--mantine-color-dark-blue-filled);--mantine-color-dark-blue-filled: var(--mantine-color-dark-blue-6);--mantine-color-dark-blue-filled-hover: var(--mantine-color-dark-blue-7);--mantine-color-dark-blue-light: rgba(5, 19, 33, 0.1);--mantine-color-dark-blue-light-hover: rgba(5, 19, 33, 0.12);--mantine-color-dark-blue-light-color: var(--mantine-color-dark-blue-6);--mantine-color-dark-blue-outline: var(--mantine-color-dark-blue-6);--mantine-color-dark-blue-outline-hover: rgba(5, 19, 33, 0.05);--mantine-color-beige-text: var(--mantine-color-beige-filled);--mantine-color-beige-filled: var(--mantine-color-beige-6);--mantine-color-beige-filled-hover: var(--mantine-color-beige-7);--mantine-color-beige-light: rgba(226, 209, 181, 0.1);--mantine-color-beige-light-hover: rgba(226, 209, 181, 0.12);--mantine-color-beige-light-color: var(--mantine-color-beige-6);--mantine-color-beige-outline: var(--mantine-color-beige-6);--mantine-color-beige-outline-hover: rgba(226, 209, 181, 0.05);}
  </style>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 29.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 30em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 63.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 64em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 89.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 90em) {.mantine-hidden-from-xl {display: none !important;}}@media (max-width: 23.99375em) {.mantine-visible-from-xxs {display: none !important;}}@media (min-width: 24em) {.mantine-hidden-from-xxs {display: none !important;}}
  </style>
  <div
    class="_wrapper_979ed7"
  >
    <div
      class="_stat_979ed7 m_7485cace mantine-Container-root"
    >
      <p
        class="mantine-focus-auto _count_979ed7 m_b6d8b162 mantine-Text-root"
      >
        1 000
      </p>
      <p
        class="mantine-focus-auto _title_979ed7 m_b6d8b162 mantine-Text-root"
      >
        Books in our collection
      </p>
      <p
        class="mantine-focus-auto _description_979ed7 m_b6d8b162 mantine-Text-root"
      >
        Explore our vast collection of books, where every title opens the door to new adventures and ideas!
      </p>
    </div>
    <div
      class="_stat_979ed7 m_7485cace mantine-Container-root"
    >
      <p
        class="mantine-focus-auto _count_979ed7 m_b6d8b162 mantine-Text-root"
      >
        200
      </p>
      <p
        class="mantine-focus-auto _title_979ed7 m_b6d8b162 mantine-Text-root"
      >
        Authors featured
      </p>
      <p
        class="mantine-focus-auto _description_979ed7 m_b6d8b162 mantine-Text-root"
      >
        Immerse yourself in stories from a diverse set of authors, each bringing a unique perspective to our library.
      </p>
    </div>
    <div
      class="_stat_979ed7 m_7485cace mantine-Container-root"
    >
      <p
        class="mantine-focus-auto _count_979ed7 m_b6d8b162 mantine-Text-root"
      >
        5 000
      </p>
      <p
        class="mantine-focus-auto _title_979ed7 m_b6d8b162 mantine-Text-root"
      >
        Number of ratings
      </p>
      <p
        class="mantine-focus-auto _description_979ed7 m_b6d8b162 mantine-Text-root"
      >
        With a massive collection of ratings, you have the insights you need to discover your next must-read.
      </p>
    </div>
  </div>
</DocumentFragment>
`;
